src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,110 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double MULTIPLIER = 1.5;
+    private static final int DEFAULT_SIZE = 10;
+    private static final int ZERO = 0;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_SIZE];
+    }
+
+    private void checkCapacity() {
+        if (size == elements.length) {
+            grow();
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < ZERO || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index %d out of bounds for size %d.".formatted(index, size));
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < ZERO || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index %d out of bounds for size %d.".formatted(index, size));
+        }
+    }
+
+    private void grow() {
+        int newSize = (int) (elements.length * MULTIPLIER);
+        Object[] newElements = new Object[newSize];
+        System.arraycopy(elements, ZERO, newElements, ZERO, size);
+        elements = newElements;
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkCapacity();
+        checkIndexForAdd(index);
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T deletedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return deletedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element != null && element.equals(elements[i]))
+                    || (element == null && elements[i] == null)) {
```

Comment:
```suggestion
            if (element != null && element.equals(elements[i])
                    || element == null && elements[i] == null) {
```
redundant brackets. Read about logical operator priority